### PR TITLE
Parameter names should be at least 3 characters long.  Fixes #195

### LINKED ISSUE #195: [objc] "Avoid one- and two-letter names for arguments" rule
Rule from [Apple](https://developer.apple.com/library/content/documentation/Cocoa/Conceptual/CodingGuidelines/Articles/NamingMethods.html#//apple_ref/doc/uid/20001282-BCIGIJJF) under "Method Arguments".

If we take this C# code for instance: https://github.com/mono/Embeddinator-4000/blob/objc/tests/managed/methods.cs#L93-L94 (and once https://github.com/mono/Embeddinator-4000/issues/193 and  https://github.com/mono/Embeddinator-4000/issues/194 are fixed :P) we should get:

```
- (int)doItWithInt32Value:(int) anInt;
- (int)doItWithStringValue:(NSString *) aString;
```

### PR DESCRIPTION
**Pull request by timrisi**

### CODE REVIEW

File: objcgen/objcgenerator-helpers.cs [hunk 1]
```diff
@@ -40,8 +76,17 @@ void GetSignatures (string objName, string monoName, MemberInfo info, ParameterI
 				var ptname = NameGenerator.GetTypeName (p.ParameterType);
 				if (types.Contains (pt))
 					ptname += " *";
+				string pName = p.Name;
+				if (p.Name.Length < 3) {
+					if (ObjCTypeToArgument.ContainsKey (ptname))
+						pName = ObjCTypeToArgument [ptname];
+					else pName = "anObject";
```
Comment (spouliot) [inline objcgen/objcgenerator-helpers.cs]: minor/style: `else` should be alone on its line

Comment (spouliot) [inline objcgen/objcgenerator-helpers.cs]: You're doing 2 lookups, why not
```
if (!ObjCTypeToArgument.TryGetValue (ptname, out pName))
    pName = "anObject";
```

File: objcgen/objcgenerator.cs [hunk 2]
```diff
@@ -469,6 +469,14 @@ void Generate (ParameterInfo [] parameters, bool isExtension, out string postInv
 			for (int i = 0; i < pcount; i++) {
 				var p = parameters [i];
 				var name = (isExtension && (i == 0)) ? "self" : p.Name;
+				if (name.Length< 3) {
+					if (CsharpTypeToArgument.ContainsKey (p.ParameterType.ToString ()))
+						name = CsharpTypeToArgument [p.ParameterType.ToString ()];
+					else name = "anObject";
```
Comment (spouliot) [inline objcgen/objcgenerator.cs]: same as above (two lookups)
also `p.ParameterType.ToString ()` is done twice

File: objcgen/objcgenerator.cs [hunk 3]
```diff
@@ -469,6 +469,14 @@ void Generate (ParameterInfo [] parameters, bool isExtension, out string postInv
...
+
+					if (parameters.Count (p2 => p2.ParameterType == p.ParameterType && p2.Name.Length< 3) > 1)
+						name += p.Name.PascalCase();
```
Comment (spouliot) [inline objcgen/objcgenerator.cs]: minor/style: `PascalCase ()`

File: objcgen/objcgenerator-helpers.cs [hunk 4]
```diff
@@ -40,8 +76,17 @@ void GetSignatures (string objName, string monoName, MemberInfo info, ParameterI
...
+
+					if (parameters.Count (p2 => p2.ParameterType == p.ParameterType && p2.Name.Length < 3) > 1)
+						pName += p.Name.PascalCase();
```
Comment (spouliot) [inline objcgen/objcgenerator-helpers.cs]: minor/style: `PascalCase ()`

**Comment by timrisi** [comment]
I'm working on tests at the moment.  I just wanted to get the PR out there for feedback on the code.  Currently for testing I'm creating C# methods using short parmater names to make sure the method signature and their use in the method itself (which I almost forgot to change before doing the PR) match and the method still works correctly after renaming.

File: objcgen/objcgenerator-helpers.cs (hunk 1 above)
Comment (timrisi) [inline objcgen/objcgenerator-helpers.cs]: Good call

File: tests/objc-cli/libmanaged/Tests/Tests.m [hunk 5]
```diff
@@ -740,6 +740,23 @@ - (void) testOperatorOverloading {
...
+    XCTAssert ([testClass twoCharQ:'c' r:'d'] == 199, "Two characters");
+    XCTAssert ([testClass twoLongM:-3 n:15] == 12, "Two longs");
+    XCTAssert ([testClass twoUintG:3 h:4] == 7, "Two uints");
+    NSLog (@"%@: %f", @"Foo", [testClass twoFloatE:1.2f f:3.4f]);
+    XCTAssert ([testClass twoFloatE:1.2f f:3.4f] - 4.6f < 0.000001, "Two floats");
+    XCTAssert ([testClass twoShortI:1 j:4] == 5, "Two shorts");
+    XCTAssert ([testClass twoUlongO:2 p:3] == 5, "Two ulong");
+    XCTAssert ([testClass twoDoubleC:3.2 d:5.7] == 8.9, "Two doubles");
+    XCTAssert ([[testClass twoStringA:@"Foo" b:@"Bar"] isEqualToString:@"FooBar"], "Two strings");
+    XCTAssert ([testClass twoUshortK:5 l:8] == 13, "Two ushorts");
```
Comment (spouliot) [inline tests/objc-cli/libmanaged/Tests/Tests.m]: Please use tabs (not spaces) for indentation

File: tests/managed/shortParameters.cs [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+            R = r;
+            return q + r;
+        }
+
+        public bool TwoBool (bool s, bool t)
+        {
+            S = s;
+            T = t;
+            return s || t;
+        }
```
Comment (spouliot) [inline tests/managed/shortParameters.cs]: mixed spaces and tabs indentation in file (if you just moved to VSfM then you need to set your preferences to Mono style)

File: objcgen/extensions.cs [hunk 7]
```diff
@@ -117,6 +119,24 @@ public static bool HasCustomAttribute (this MemberInfo self, string @namespace,
...
 	}
 
+	public static class ParameterInfoExtensions {
+		public static string ExtendedName (this ParameterInfo p, ParameterInfo[] parameters)
+		{
+			string pName = p.Name;
+			string ptname = NameGenerator.GetTypeName (p.ParameterType);
+			if (p.Name.Length< 3) {
+				if (!ObjCGenerator.ObjCTypeToArgument.TryGetValue(ptname, out pName))
+                    pName = "anObject";
```
Comment (spouliot) [inline objcgen/extensions.cs]: indent issue (spaces vs tabs)

File: tests/objc-cli/libmanaged/Tests/Tests.m (hunk 5 above)
Comment (timrisi) [inline tests/objc-cli/libmanaged/Tests/Tests.m]: Whoops, apparnetly my xcode was still defaulted to doing spaces for indentation

**Comment by spouliot** [comment]
Please double check the build failure before merging

File: objcgen/extensions.cs [hunk 8]
```diff
@@ -117,6 +119,24 @@ public static bool HasCustomAttribute (this MemberInfo self, string @namespace,
 		}
 	}
 
+	public static class ParameterInfoExtensions {
+		public static string ExtendedName (this ParameterInfo p, ParameterInfo[] parameters)
+		{
+			string pName = p.Name;
+			string ptname = NameGenerator.GetTypeName (p.ParameterType);
+			if (p.Name.Length< 3) {
```
Comment (chamons) [inline objcgen/extensions.cs]: space before <

File: objcgen/objcgenerator-helpers.cs [hunk 9]
```diff
@@ -12,6 +12,29 @@ public partial class ObjCGenerator {
 
 		// get a name that is safe to use from ObjC code
 
+		public static Dictionary<string, string> ObjCTypeToArgument = new Dictionary<string, string> {
```
Comment (chamons) [inline objcgen/objcgenerator-helpers.cs]: Part of me hates to add _more_ code to any of the objcgenerator* files.

@spouliot @timrisi - Should we hoist all of this "name geneartion code to the name generator?

Comment (spouliot) [inline objcgen/objcgenerator-helpers.cs]: I don't think this was merged when the PR go out initially but yes it makes sense, now, to move the code there.

Comment (timrisi) [inline objcgen/objcgenerator-helpers.cs]: I'll go take a look

**Comment by timrisi** [comment]
Created an issue for the build failure: https://github.com/mono/Embeddinator-4000/issues/230

**Comment by spouliot** [comment]
Build failure fixed by PR #231

**Comment by spouliot** [comment]
build

**Comment by timrisi** [comment]
@chamons Need updated review

File: objcgen/extensions.cs [hunk 10]
```diff
@@ -117,6 +119,24 @@ public static bool HasCustomAttribute (this MemberInfo self, string @namespace,
 		}
 	}
 
+	public static class ParameterInfoExtensions {
+		public static string ExtendedName (this ParameterInfo p, ParameterInfo[] parameters)
```
Comment (chamons) [inline objcgen/extensions.cs]: This should stop being an extension method and just a regular method on namegenerator.

**Pull request merged**